### PR TITLE
music.py

### LINKED ISSUE #1: Combine the play and play_from commands
Hi! This is a [`first-timers-only`](https://www.firsttimersonly.com/) issue.

This means it's aimed towards people who haven't contributed to open source before, or are just starting to. As such, it is a relatively simple issue, so people can focus more on the process of contributing rather than the complexity of the issue. 

If that's you, keep reading to find more about the problem, the expected solution, and any proposed steps towards that solution. 

## The problem

Currently there are two commands to start playing music: `play` and `play_from`. `play` takes a query (name of an artist, of a song, etc) as an argument and searches YouTube to get a result, while `play_from` takes a URL directly. For ease of use, only one command should be needed to deal with both cases.

## The solution

As it turns out, the feature is almost there: `play` can already handle both cases! Searching a video's URL in YouTube returns the same video (as it should!).

So an easy solution is to combine both commands into `play`, and use it to search both queries and URLs.

A better solution would be to detect whether the provided argument is a URL. That way, we could pass it directly to the function that actually gets the song instead of searching YouTube to get the same URL first.

## How to implement it

Basically, the steps to do this are:

- Claim this issue by commenting below. If someone else has claimed it, ask if they've opened a pull request already and if they're stuck -- maybe you can help them solve a problem or move it along!
- Take a look at CONTRIBUTING for guidelines, and at README for some context.
- Clone the repo if you haven't already.
- Create a new branch, preferably with a descriptive such as `combine_play`.
- Go to `cogs/music.py`.
- Currently the `play` command starts at [line 206](https://github.com/JulioLoayzaM/CroissantBot/blob/main/cogs/music.py#L206) and `play_from` starts at [line 246](https://github.com/JulioLoayzaM/CroissantBot/blob/main/cogs/music.py#L246). These are the two functions to modify.
- Combine the two:
  - The new `play` function should have three parameters: the two necessary ones, `self` and `ctx`, and a `query` parameter.
  - It should use the code from the current `play` function to perform the search, and the code from `play_from` to use the new URL as it currently does: in other words, copy the code from `play` and paste it after the `await msg.edit(suppress=True)` at [line 267](https://github.com/JulioLoayzaM/CroissantBot/blob/main/cogs/music.py#L267).
- That should be it! You can create a pull request and wait for review. If you can test the changes beforehand, please do it.

In the meantime, there are a couple of bonus points you can tackle (no pressure!):
- Update the docstring, type-hints and command help to match the new function.
- Now that you should have only one `play` function, people that were using `play_from` (if any) might wonder why the bot suddenly doesn't recognize the command. To avoid this, create a new `play_from` function. This new function should be quite short:
  - It should notify the user that the command is now deprecated. You can send a message with `await ctx.send(message)`.
  - You can also add a notification by changing the help text above the function.
  - It should still take a URL as argument, but pass it immediatly to `play` with `await self.play(ctx, url)` instead of processing it first.
  - That way the `play_from` command still exists and works, but users are made aware it will go away in a future version.
- Finally, you can try to implement the better solution mentionned above: instead of blindly searching for whatever query it gets, `play` should try to recognize if the argument is a URL. If it is one, it can pass it directly to `YTDLSource.from_url` instead of searching for a result first.

That's it. :)

I said it is a simple issue, but it can be fairly overwhelming if you don't have much or any experience with `git`, `discord.py`, or even the `asyncio` library. Feel free to ask for help at any point, even (or specially) at the beginning. Remember, the idea of this issue is to help people get the hang of contributing to projects; the problem itself is just an entertaining mean to that end. As mentionned, there's also no pressure on doing the bonus points.

One last thing: I'd appreciate to have some update on your progress, at most a week after making the claim, even if it's to say you're still working on it. That way I can know whether to maintain your assignment or if I should assign it to someone else.

### PR DESCRIPTION
**Pull request by TroySherlock**
## Summary

Merged play and play_for functions. Renamed to play(self, ctx: commands.Context, query). Removed 'if url ==none:' statement.

## Checklist

- [x] If code changes were made, they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, etc)

### CODE REVIEW

**Comment by JulioLoayzaM** [comment]
Hi @TroySherlock,

Thanks for submitting a pull request! I'm currently going through the code and I'll add some comments.

Just wanted to point out to that it's good practice to comment on an issue to inform others that you're working on it. That way, they can know that someone is already taking care of the issue.

**Comment by TroySherlock** [comment]
@JulioLoayzaM My bad I didn't know. Do they want to try the rest? I can leave the rest of the issue, if that works.

**Comment by JulioLoayzaM** [comment]
Don't worry, I think you're almost done with it anyway and I was already planning to open another issue.

File: cogs/music.py [hunk 1]
```diff
@@ -235,6 +235,10 @@ async def play(self, ctx: commands.Context, query):
 		Parameters:
 			- query: the query to search for in youtube.
 		"""
+		if query == None:
+			await ctx.send(f"You have to provide a youtube url or query. Type `{BOT_PREFIX}play <url/query>.")
+			return
```
Comment (JulioLoayzaM) [inline cogs/music.py:242]: I like the return better. :)

File: cogs/music.py [hunk 2]
```diff
@@ -235,6 +235,10 @@ async def play(self, ctx: commands.Context, query):
 		Parameters:
 			- query: the query to search for in youtube.
 		"""
+		if query == None:
```
Comment (JulioLoayzaM) [inline cogs/music.py]: Just one little detail, it's better to use `query is None` rather than `query == None`. Same goes for `play_from`.

**Review by JulioLoayzaM — CHANGES_REQUESTED** [review]

**Comment by JulioLoayzaM** [comment]
The two cases of `query == None` are the only changes left, otherwise it's good to go!

**Comment by JulioLoayzaM** [comment]
Oh and you could edit the README to mark the task as done. ;)

**Comment by JulioLoayzaM** [comment]
Hi @TroySherlock,

I haven't missed your commits! The first two are good to go. I was checking the third one, I'll take a better look in the morning; I'm not familiar with youtube-dl's extractors.

File: cogs/music.py [hunk 3]
```diff
@@ -239,15 +241,17 @@ async def play(self, ctx: commands.Context, query: str=None):
...
 		# To avoid clutter, we edit the user's message to suppress the embed
 		msg = ctx.message
 		await msg.edit(suppress=True)
 
-		# to use search instead of passing an url, prepend 'ytsearch:' to the search terms
-		# to search for more than one video, prepend 'ytsearchx:' with x being the number of videos
-		video = ytdl.extract_info(f"ytsearch:{query}", download=False)['entries'][0]
-		url = video['webpage_url']
+		# Checks if query is a valid url, if not we search youtube for the query
+		if validate_url(query) is False:
```
Comment (JulioLoayzaM) [inline cogs/music.py]: Since `validate_url` returns a bool, the test should be `if not await validate_url(query)`.

File: cogs/music.py [hunk 4]
```diff
@@ -966,6 +970,25 @@ async def get_song_info(self) -> str:
 		return f"{self.song.title} - {self.song.url}"
 
 
+def validate_url(url):
```
Comment (JulioLoayzaM) [inline cogs/music.py]: It's better to make non-critical functions `async`, then `await` the result.

File: cogs/music.py [hunk 5]
```diff
@@ -966,6 +970,25 @@ async def get_song_info(self) -> str:
...
+	Checks to see if url has any valid extractors for youtube_dl
+
+	Parameters:
+		- url: the url to search for extractors.
+
+	Returns:
+		-True, if site has dedicated extractor
+		-False, if site has no dedicated extractor
+	"""
+	extractors = youtube_dl.extractor.gen_extractors()
```
Comment (JulioLoayzaM) [inline cogs/music.py]: We're expecting a youtube link, so we can get the youtube extractor directly with `youtube_dl.extractor.get_info_extractor('Youtube')`. Then, we can just `return e.suitable(url)`.

**Review by JulioLoayzaM — CHANGES_REQUESTED** [review]
I didn't know about extractors, this solves the issue of checking for a URL. Nice find! :)

There are a couple of changes to improve `validate_url` and we're done.

File: cogs/music.py [hunk 6]
```diff
@@ -970,23 +970,19 @@ async def get_song_info(self) -> str:
...
 	"""
 	Checks to see if url has any valid extractors for youtube_dl
 
 	Parameters:
 		- url: the url to search for extractors.
 
 	Returns:
-		-True, if site has dedicated extractor
-		-False, if site has no dedicated extractor
+		-extractor, if site has dedicated extractor
```
Comment (JulioLoayzaM) [inline cogs/music.py]: It returns a bool (True/False) rather than an extractor.

**Review by JulioLoayzaM — CHANGES_REQUESTED** [review]
One small detail in the dosctring before merging.

File: cogs/music.py (hunk 6 above)
Comment (TroySherlock) [reply cogs/music.py:981]: Are you sure? I thought it returns an object, but it evaluates to true.

Comment (TroySherlock) [reply cogs/music.py:981]: I might have misread tho.

Comment (TroySherlock) [reply cogs/music.py]: Actually nevermind I think your right.

**Review by JulioLoayzaM — APPROVED** [review]
That should be it!

**Comment by JulioLoayzaM** [comment]
Done! Thank you for your contribution @TroySherlock !

**Pull request merged**